libsplinter/src/auth/identity/mod.rs

```diff
@@ -0,0 +1,83 @@
+// Copyright 2018-2020 Cargill Incorporated
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Tools for identifying clients and users
+
+mod error;
+
+pub use error::IdentityProviderError;
+
+/// A service that fetches identities from a backing provider
+pub trait IdentityProvider: Send {
+    /// Attempts to get the identity that corresponds to the given credentials
+    fn get_identity(&self, credentials: &AuthCredentials) -> Result<String, IdentityProviderError>;
+
+    /// Clones implementation for `IdentityProvider`. The implementation of the `Clone` trait for
+    /// `Box<dyn IdentityProvider>` calls this method.
+    ///
+    /// # Example
+    ///
+    ///```ignore
+    ///  fn clone_box(&self) -> Box<dyn IdentityProvider> {
+    ///     Box::new(self.clone())
+    ///  }
+    ///```
+    fn clone_box(&self) -> Box<dyn IdentityProvider>;
+}
+
+impl Clone for Box<dyn IdentityProvider> {
+    fn clone(&self) -> Box<dyn IdentityProvider> {
+        self.clone_box()
+    }
+}
+
+/// The authentication credentials that are passed to an `IdentityProvider`
+pub struct AuthCredentials {
+    /// The authentication type. Each identity provider may support only certain types.
+    auth_type: String,
```

Comment:
Should this be an enum?

Reply:
That was my initial thought, but that would restrict the API to only support types implemented by the library itself; you couldn't add external implementations for custom types without modifying the library. Though, we could have an enum with known types, and a `Custom(String)` variant.

Reply:
I think that would be a good compromise

Reply:
Replaced this struct with a couple enums that provide better structure and typing to the API.